crypto/err/err_blocks.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/err.h>
+#include "err_locl.h"
+
+void ERR_new(void)
+{
+    ERR_STATE *es;
+
+    es = ERR_get_state();
+    if (es == NULL)
+        return;
+
+    /* Allocate a slot */
+    err_allocate(es);
+    err_clear(es, es->top);
+}
+
+void ERR_set_debug(const char *file, int line, const char *func)
+{
+    ERR_STATE *es;
+
+    es = ERR_get_state();
+    if (es == NULL)
+        return;
+
+    err_set_debug(es, es->top, file, line, func);
+}
+
+void ERR_set_error(int lib, int reason, const char *fmt, ...)
+{
+    va_list args;
+
+    va_start(args, fmt);
+    ERR_vset_error(lib, reason, fmt, args);
+    va_end(args);
+}
+
+void ERR_vset_error(int lib, int reason, const char *fmt, va_list args)
+{
+    ERR_STATE *es;
+    char *buf = NULL;
+    size_t buf_len = 0;
+
+    es = ERR_get_state();
+    if (es == NULL)
+        return;
+
+    if (fmt != NULL
+        && (buf = OPENSSL_malloc(ERR_MAX_DATA_SIZE)) != NULL) {
```

Comment:
This approach, alloc potential worst-case then shrink it, has two failure points. If malloc fails use fixed-string like "extra data suppressed, allocation failure" or some such. If you are that concerned about the size, call vsnprintf twice, or start at 1K and go up.

Reply:
I've reduced the size to 1K

Reply:
I still think that is way to big and suggest 256 (sorry 1K looked big compared to 4K); four lines of typical text are still way more than we need to support.  Don't realloc, let vsnprintf do its truncation work.  And keep/re-use the pointer.